modules/related-posts/jetpack-related-posts.php

```diff
@@ -246,6 +251,118 @@ public function get_target_html_unsupported() {
 		return "\n\n<!-- Jetpack Related Posts is not supported in this context. -->\n\n";
 	}
 
+	/**
+	 * ===============
+	 * GUTENBERG BLOCK
+	 * ===============
+	 */
+
+	/**
+	 * Register the Related Posts Gutenberg block on server.
+	 *
+	 * @action init
+	 * @uses register_block_type
+	 * @return null
+	 */
+	public function register_block() {
+		register_block_type(
+			'a8c/related-posts',
+			array(
+				'render_callback' => array( $this, 'render_block' ),
+			)
+		);
+	}
+
+	/**
+	 * Render the related posts markup.
+	 *
+	 * @return string
+	 */
+	public function render_block( $attributes ) {
+		$block_attributes = array(
+			'show_thumbnails' => isset( $attributes['displayThumbnails'] ) ? (bool) $attributes['displayThumbnails'] : false,
+			'show_date' => isset( $attributes['displayDate'] ) ? (bool) $attributes['displayDate'] : true,
+			'show_context' => isset( $attributes['displayContext'] ) ? (bool) $attributes['displayContext'] : true,
+			'layout' => isset( $attributes['postLayout'] ) && $attributes['postLayout'] === 'list' ? $attributes['postLayout'] : 'grid',
+			'size' => ! empty( $attributes['postsToShow'] ) ? absint( $attributes['postsToShow'] ) : 3,
+		);
+
+		$related_posts = $this->get_for_post_id( get_the_ID(), array(
+			'size' => $block_attributes['size'],
+		) );
+
+		if ( ! $related_posts ) {
+			return '';
+		}
+
+		ob_start();
+		?>
+		<div id="jp-relatedposts" class="jp-relatedposts" style="display: block;">
```

Comment:
I am seeing 2 duplicate sets of 3 related posts on my posts. 

<img width="554" alt="screenshot 2018-10-18 at 13 05 25" src="https://user-images.githubusercontent.com/426388/47150318-828e5380-d2d6-11e8-8eaa-4c1f8eacd542.png">

One is added via the Gutenberg block, the other is added via the regular Related Posts and gets injected right here because it finds the `jp-relatedposts` ID in the post content:
https://github.com/Automattic/jetpack/blob/bb884c04b32a57e3053f14c56e2f4e0138aef185/modules/related-posts/related-posts.js#L231

What should the expected behaviour be here? Should we hide the posts at the bottom of the page when a block has been added in the post content, like we do with the Related Posts shortcode?
https://github.com/Automattic/jetpack/blob/bb884c04b32a57e3053f14c56e2f4e0138aef185/modules/related-posts/jetpack-related-posts.php#L171

Or should we display the same Related Posts twice, once in the post content, once below it? If we choose to go that way, we should make sure the Gutenberg Related Posts don't use the same ID.

Reply:
If we're going to allow multiple blocks like this, we can't use a static ID for them, that would be invalid HTML. IDs must be globally unique in the document.

See https://github.com/Automattic/wp-calypso/pull/27927#issuecomment-431044923

That would solve the some of the problems @jeherve [mentions](https://github.com/Automattic/jetpack/pull/10132/files#r226263162), but unfortunately requires a fair amount of styling work, as this is used for lots of styling: https://github.com/Automattic/jetpack/search?l=CSS&q=jp-relatedposts

Reply:
So, if we want to reuse most of the existing functionality (which I guess we do), we should keep using the same ID, but also to have a maximum of 1 of those blocks per post. So I've suggested a PR to add this limitation to the block: 

https://github.com/Automattic/wp-calypso/pull/29415

Additionally, some small modifications are necessary so the related posts script will be able to properly work, but not load additional posts on top of ours. I've done that here: https://github.com/Automattic/jetpack/pull/10132/commits/73b1f29b3c01fa2845b75b14070eaf264cbbabeb